openslides/agenda/views.py

```diff
@@ -811,3 +814,69 @@ def get_queryset(self):
         if not self.request.user.has_perm('agenda.can_see_orga_items'):
             queryset = queryset.exclude(type__exact=Item.ORGANIZATIONAL_ITEM)
         return queryset
+
+    @list_route(methods=['get', 'put'])
+    def tree(self, request):
+        """
+        Returns or set the agenda tree.
+        """
+        if request.method == 'PUT':
+            return self.set_tree(request)
+        return self.get_tree(request)
+
+    def get_tree(self, request):
+        """
+        Returns the agenda tree.
+        """
+        item_list = Item.objects.order_by('weight')
+
+        # Index the items to get the children for each item
+        item_children = defaultdict(list)
+        for item in item_list:
+            if item.parent:
+                item_children[item.parent_id].append(item)
+
+        def get_children(item):
+            """
+            Returns a list with all the children for item.
+
+            Returns an empty list if item has no children.
+            """
+            return [dict(id=child.pk, children=get_children(child))
+                    for child in item_children[item.pk]]
+
+        return Response(dict(id=item.pk, children=get_children(item))
+                        for item in item_list if not item.parent)
+
+    def set_tree(self, request):
+        """
+        Sets the agenda tree.
+        """
+        tree = request.data['tree']
+
+        def walk_items(tree, parent=None):
+            """
+            Generator that returns each item in the tree as tuple.
+
+            This tuples have tree values. The item id, the item parent and the
+            weight of the item.
+            """
+            for weight, element in enumerate(tree):
+                yield (element['id'], parent, weight)
+                yield from walk_items(element['children'], element['id'])
+
+        touched_items = set()
+        for item_pk, parent_pk, weight in walk_items(tree):
+            # Check that the item is only once in the tree to prevent invalid trees
+            if item_pk in touched_items:
+                status = "Error: Item %d is more then once in the tree" % item_pk
+                break
+            touched_items.add(item_pk)
+
+            Item.objects.filter(pk=item_pk).update(
+                parent_id=parent_pk,
+                weight=weight)
+        else:
+            status = "ok"
+
+        return Response({'status': status})
```

Comment:
On every other API request the framework does so. HTTP 200 means OK and HTTP 400 means Error. Additionally the framework uses a 'detail' message. So we should do this as the framework does.